tests/MongoDB.Driver.Tests/Specifications/auth/AuthTestRunner.cs

```diff
@@ -0,0 +1,147 @@
+        {
+            var definition = testCase.Test;
+            JsonDrivenHelper.EnsureAllFieldsAreValid(definition, "description", "uri", "valid", "credential");
+
+            MongoCredential mongoCredential = null;
+            Exception parseException = null;
+            try
+            {
+                var connectionString = (string)definition["uri"];
+                mongoCredential = MongoClientSettings.FromConnectionString(connectionString).Credential;
+            }
+            catch (Exception ex)
+            {
+                parseException = ex;
+            }
+
+            if (parseException == null)
+            {
+                AssertValid(mongoCredential, definition);
+            }
+            else
+            {
+                AssertInvalid(parseException, definition);
+            }
+        }
+
+        private void AssertValid(MongoCredential mongoCredential, BsonDocument definition)
+        {
+            if (!definition["valid"].ToBoolean())
+            {
+                throw new AssertionException($"The connection string '{definition["uri"]}' should be invalid.");
+            }
+
+            var credential = definition["credential"] as BsonDocument;
+            if (credential != null)
+            {
+                JsonDrivenHelper.EnsureAllFieldsAreValid(credential, "username", "password", "source", "mechanism", "mechanism_properties");
+                mongoCredential.Username.Should().Be(ValueToString(credential["username"]));
+#pragma warning disable 618
+                mongoCredential.Password.Should().Be(ValueToString(credential["password"]));
+#pragma warning restore 618
+                mongoCredential.Source.Should().Be(ValueToString(credential["source"]));
+                mongoCredential.Mechanism.Should().Be(ValueToString(credential["mechanism"]));
+
+                var authenticator = mongoCredential.ToAuthenticator();
+                if (credential.TryGetValue("mechanism_properties", out var mechanismProperties))
+                {
+                    if (mechanismProperties.IsBsonNull)
+                    {
+                        if (authenticator is GssapiAuthenticator)
+                        {
+                            var serviceName = authenticator._mechanism()._serviceName();
+                            serviceName.Should().Be("mongodb"); // The default is "mongodb".
+                            var canonicalizeHostName = authenticator._mechanism()._canonicalizeHostName();
+                            canonicalizeHostName.Should().BeFalse(); // The default is "false".
+                        }
+                        else
+                        {
+                            // Other authenticators do not contain mechanism properties
+                        }
+                    }
+                    else
+                    {
+                        foreach (var mechanismProperty in mechanismProperties.AsBsonDocument)
```

Comment:
Consider renaming `mechanismProperty` to `expectedMechanismProperty`.

Reply:
Done